core/consensus/validation/babe_block_validator.cpp

```diff
@@ -0,0 +1,222 @@
+
+    // secondly, retrieve public key of the peer by its authority id
+    if (static_cast<uint64_t>(authorities.size())
+        < babe_header.authority_index) {
+      log_->info("don't know about authority with index {}",
+                 babe_header.authority_index);
+      return false;
+    }
+    const auto &key = authorities[babe_header.authority_index].id;
+
+    // thirdly, use verify function to check the signature
+    return sr25519_verify(seal.signature.data(),
+                          block_hash.data(),
+                          decltype(block_hash)::size(),
+                          key.data());
+  }
+
+  bool BabeBlockValidator::verifyVRF(const BabeBlockHeader &babe_header,
+                                     gsl::span<const Authority> authorities,
+                                     const Randomness &randomness,
+                                     const Threshold &threshold) const {
+    // verify VRF output
+    auto randomness_with_slot = Buffer{}
+                                    .putBytes(uint256_t_to_bytes(randomness))
+                                    .putBytes(uint128_t_to_bytes(threshold));
+    if (!vrf_provider_->verify(randomness_with_slot,
+                               babe_header.vrf_output,
+                               authorities[babe_header.authority_index].id)) {
+      log_->info("VRF proof in block is not valid");
+      return false;
+    }
+
+    // verify threshold
+    if (babe_header.vrf_output.value >= threshold) {
+      log_->info("VRF value is not less than the threshold");
+      return false;
+    }
+
+    return true;
+  }
+
+  bool BabeBlockValidator::verifyProducer(const PeerId &peer,
+                                          BabeSlotNumber number) {
+    auto slot_it = blocks_producers_.find(number);
+    if (slot_it == blocks_producers_.end()) {
+      // this peer is the first producer in this slot
+      blocks_producers_.insert({number, std::unordered_set<PeerId>{peer}});
+      return true;
+    }
+
+    auto &slot = slot_it->second;
+    auto peer_in_slot = slot.find(peer);
+    if (peer_in_slot != slot.end()) {
+      log_->info("peer {} has already produced a block in the slot {}",
+                 peer.toBase58(),
+                 number);
+      return false;
+    }
+
+    // OK
+    slot.insert(peer);
+    return true;
+  }
+
+  bool BabeBlockValidator::verifyTransactions(
+      const primitives::BlockBody block_body) const {
+    return std::all_of(
+        block_body.cbegin(), block_body.cend(), [this](const auto &ext) {
+          auto validation_res = tx_queue_->validate_transaction(ext);
+          if (!validation_res) {
+            log_->info("extrinsic validation failed: {}",
+                       validation_res.error());
+            return false;
+          }
+          return visit_in_place(
+              validation_res.value(),
+              [](primitives::Valid) { return true; },
+              [](auto) { return false; });
```

Comment:
You sure it will work stable? Maybe it would be more reliable to just explicitly put other possible values, given that there's only two of them?